### PR TITLE
Add swagger editor/tryit for API and API generation

### PR DESCRIPTION
**Pull request by milindaperera**
## Purpose
Add swagger editor and API generate support
Fixes: wso2/product-ei/issues/3328
    

- Edit (swagger ui [2]) and publish a swagger definition

Add an embedded swagger editor for the API resources created in WSO2 EI. The swagger document is generated for the API and stored as a registry resource. The generated swagger document can be modified with the swagger editor.
- Try-it support for APIs using swagger editor
- API generation (skeleton of API) from the swagger definition.

### CODE REVIEW

File: components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/service/RestApiAdmin.java [hunk 1]
```diff
@@ -928,5 +958,173 @@ protected ConfigurationContext getConfigContext() {
...
+    /**
+     * Function to generate API from swagger definition (from JSON representation)
+     *
+     * @param swaggerJsonString swagger definition
+     * @return generated synapse API
+     * @throws APIException
+     */
+    public String generateAPIFromSwagger(String swaggerJsonString) throws APIException {
+
+        if (swaggerJsonString == null || swaggerJsonString.isEmpty()) {
```
Comment (sameeragunarathne) [inline components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/service/RestApiAdmin.java]: AFAIR we can use StringUtils.isEmpty()

File: components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/service/RestApiAdmin.java [hunk 2]
```diff
@@ -928,5 +958,173 @@ protected ConfigurationContext getConfigContext() {
...
+        }
+
+        JsonParser jsonParser = new JsonParser();
+        JsonElement swaggerJson = jsonParser.parse(swaggerJsonString);
+        if (swaggerJson.isJsonObject()) {
+            APIGenerator apiGenerator = new APIGenerator(swaggerJson.getAsJsonObject());
+            try {
+                API api = apiGenerator.generateSynapseAPI();
+                return APISerializer.serializeAPI(api).toString();
+            } catch (SwaggerGenException e) {
```
Comment (malakaganga) [inline components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/service/RestApiAdmin.java]: Shall we see whether it is required to log and throw the exception here?

File: components/org.wso2.carbon.integrator.core/src/main/java/org/wso2/carbon/integrator/core/rest/api/swagger/GenericApiObjectDefinition.java [hunk 3]
```diff
@@ -0,0 +1,410 @@
+/*
+ * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (madushadhanushka) [inline components/org.wso2.carbon.integrator.core/src/main/java/org/wso2/carbon/integrator/core/rest/api/swagger/GenericApiObjectDefinition.java]: 2016 license header

Comment (dulanjalidilmi) [inline components/org.wso2.carbon.integrator.core/src/main/java/org/wso2/carbon/integrator/core/rest/api/swagger/GenericApiObjectDefinition.java]: Shall we add correct year?

File: components/mediation-ui/org.wso2.carbon.rest.api.ui/src/main/resources/web/api/generateAPIWizard2.jsp [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    ConfigurationContext configContext = (ConfigurationContext) config
+            .getServletContext().getAttribute(
+                    CarbonConstants.CONFIGURATION_CONTEXT);
+    String cookie = (String) session
+            .getAttribute(ServerConstants.ADMIN_SERVICE_COOKIE);
+    RestApiAdminClient client = new RestApiAdminClient(
+            configContext, url, cookie, bundle.getLocale());
+
+    String apiName = "";
+    String apiContext = "";
```
Comment (dulanjalidilmi) [inline components/mediation-ui/org.wso2.carbon.rest.api.ui/src/main/resources/web/api/generateAPIWizard2.jsp]: Shall we remove unused variables?

File: components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/service/RestApiAdmin.java (hunk 1 above)
Comment (milindaperera) [reply components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/service/RestApiAdmin.java]: We have to depend on a dependency just for this check. Hence shall we avoid adding a dependency for apache commons-lang

Comment (sameeragunarathne) [reply components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/service/RestApiAdmin.java]: +1

File: components/org.wso2.carbon.integrator.core/src/main/java/org/wso2/carbon/integrator/core/rest/api/swagger/SwaggerConstants.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+    /**
+     * Swagger element "consumes" in http://swagger.io/specification/
+     */
+    public static String CONSUMES = "consumes";
+
+    /**
+     * Swagger element "consumes" in http://swagger.io/specification/
+     */
+    public static String PRODUCES = "produces";
```
Comment (abeykoon) [inline components/org.wso2.carbon.integrator.core/src/main/java/org/wso2/carbon/integrator/core/rest/api/swagger/SwaggerConstants.java]: Comment should be corrected.

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerJsonProcessor.java [hunk 6]
```diff
@@ -41,14 +44,27 @@
...
+                responseString = retrieveFromRegistry(api, request);
+                if (responseString == null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Generating swagger definition for: " + api.getName());
+                    }
+                    JSONObject jsonDefinition = new JSONObject(new GenericApiObjectDefinition(api).getDefinitionMap());
+                    responseString = jsonDefinition.toString();
+                }
+            } catch (RegistryException e) {
+                String errMsg = "Error occurred while retrieving swagger definition from registry";
```
Comment (abeykoon) [inline components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerJsonProcessor.java]: Should we log and throw the exception?

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerYamlProcessor.java [hunk 7]
```diff
@@ -47,9 +62,119 @@ public void process(CarbonHttpRequest request, CarbonHttpResponse response,
...
         }
     }
+
+    /**
+     * Function to convert from GSON object model to generic map model.
+     *
+     * @param jsonElement gson object to transform
+     * @return
+     */
+    private Map jsonObjectToMap(JsonElement jsonElement) {
```
Comment (abeykoon) [inline components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerYamlProcessor.java]: It is better if we move utility functions to a separate class.

File: components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/service/RestApiAdmin.java (hunk 2 above)
Comment (milindaperera) [reply components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/service/RestApiAdmin.java]: Fixed, Thanks

File: components/org.wso2.carbon.integrator.core/src/main/java/org/wso2/carbon/integrator/core/rest/api/swagger/GenericApiObjectDefinition.java (hunk 3 above)
Comment (milindaperera) [reply components/org.wso2.carbon.integrator.core/src/main/java/org/wso2/carbon/integrator/core/rest/api/swagger/GenericApiObjectDefinition.java]: Fixed, Thanks

Comment (milindaperera) [reply components/org.wso2.carbon.integrator.core/src/main/java/org/wso2/carbon/integrator/core/rest/api/swagger/GenericApiObjectDefinition.java]: Fixed, Thanks

File: components/mediation-ui/org.wso2.carbon.rest.api.ui/src/main/resources/web/api/generateAPIWizard2.jsp (hunk 4 above)
Comment (milindaperera) [reply components/mediation-ui/org.wso2.carbon.rest.api.ui/src/main/resources/web/api/generateAPIWizard2.jsp]: Fixed, Thanks

File: components/org.wso2.carbon.integrator.core/src/main/java/org/wso2/carbon/integrator/core/rest/api/swagger/SwaggerConstants.java (hunk 5 above)
Comment (milindaperera) [reply components/org.wso2.carbon.integrator.core/src/main/java/org/wso2/carbon/integrator/core/rest/api/swagger/SwaggerConstants.java]: Fixed, Thanks

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerJsonProcessor.java (hunk 6 above)
Comment (milindaperera) [reply components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerJsonProcessor.java]: Fixed, Thanks

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerYamlProcessor.java (hunk 7 above)
Comment (milindaperera) [reply components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerYamlProcessor.java]: Fixed, Thanks

**Comment by milindaperera** [comment]
closing the PR since required to do few more changes

**Pull request closed without merging**